Make sure to close socket on connection timeout

puzpuzpuz opened the pull request:
Closes: #737

Connection manager was not closing the underlying connection and socket in case of connection timeout. This could potentially lead to socket leaks.

Also includes the following:
* Fix request timeout handling in `HazelcastCloudDiscovery`. The [`timeout` event](https://nodejs.org/api/http.html#http_event_timeout) wasn't handled correctly
* Some minor improvements in tests.

mdumandag commented on src/network/ClientConnectionManager.ts:
```diff
@@ -277,8 +277,15 @@ export class ClientConnectionManager extends EventEmitter {
             this.connectionTimeoutMillis,
-            new HazelcastError(`Connection timed out to address ${address}.`)
-        ).finally(() => this.pendingConnections.delete(addressKey));
+            new HazelcastError('Connection timed out to address ' + address.toString())
```
Since the connection timeout logic is moved inside the `connectXXXSocket` functions, why do we still need a `timedPromise` here?

puzpuzpuz replied:
That's because we still have other method invocations here where we do network comms without preserving the timeout. Examples of such methods are `translateAddress` (which may lead to a network call to the Hazelcast Cloud server) and `authenticateOnCluster`. That's why I decided to keep the per-`getOrConnect` timeout in place.

puzpuzpuz replied:
WDYT?

mdumandag replied:
I see, but wouldn't this be a problem for the authentication part. In there, we have 120 seconds of authentication timeout (same as the Java client), and because of this check, it might timeout earlier than that

puzpuzpuz replied:
Yes, keeping things consistent with Java client certainly makes sense. Removed the `timedPromise` part in 15279db69c82dc23ce97e5a4f0aab289e2727555

puzpuzpuz commented:
@mdumandag I've addressed your comment and fixed an issue with `HazelcastCloudDiscovery`. It wasn't handling request as it's required by `https` module (the [`timeout` event](https://nodejs.org/api/http.html#http_event_timeout) wasn't handled correctly).

Could you check the last 3 commits?

mdumandag reviewed (approved):
Thanks for the changes, looks good

puzpuzpuz commented:
@mdumandag many thanks for the review!
